text/0000-let.md

```diff
@@ -0,0 +1,265 @@
+merely through the act of expressing it.
+
+While these are advantages that a large portion of the community finds
+empowering, they do not come without drawbacks.
+
+If an expression becomes sufficiently complex, then it actually begins
+to _reduce_ clarity rather than improve it; especially when you end
+up using it more than once.
+
+Consider:
+
+``` handlebars
+Welcome back {{concat (capitalize person.firstName) ' ' (capitalize person.lastName)}}
+
+Account Details:
+First Name: {{capitalize person.firstName}}
+last Name: {{capitalize person.lastName}}
+```
+
+The moment you need to re-use values, or compose them from
+constituent values that are themselves re-used, you need to
+unroll the whole thing and move towards composition via computed properties.
+
+The `let` helper relieves this pressure by allowing you to bind the
+value of computation to a name so that it can be used to derive
+subsequent values. It _preserves the advantages of proximity and
+dependency management, while retaining the re-use and composability
+of computed properties_.
+
+With `let`, the above example can be re-written as:
+
+``` handlebars
+{{let firstName=(capitalize person.firstName) lastName=(capitalize person.lastName)}}
+
+Welcome back {{concat firstName ' ' lastName}}
+
+Account Details:
+First Name: {{firstName}}
+last Name: {{lastName}}
+```
+
+By allowing you to bind a computation to a name, the `let` helper lets
+you define computed properties right within your template.
+
+
+# Detailed design
+
+The `let` keyword is implemented as a block helper that takes
+its positional parameters and yields them on to passed block. In this
+way, it is very similar to the `with` keyword (in fact, the original
+implementation was cribbed from there). The only difference is that
+`let` will _always_ render its block. Whereas `with` will _not_
+render if it is passed an empty list, `null`, `undefined`, `false`, or
+an empty string, `let` binds all values equitably.
+
+```handlebars
+{{#with (array) as |list|}}
+  {{!does not render}}
+  You'll never see me!
+{{/with}}
+```
+
+```handlebars
+{{#let (array) as |list|}}
+  Hello there!
+{{/let}}
+```
+
+An inline form is also supported via a syntax transform, so
+
+
+``` handlebars
+{{let a=1 b=2}}
+a + b = {{sum a b}}
+```
+
+expands to
+
+``` handlebars
+{{#let 1 2 as |a b|}}
+  a + b = {{sum a b}}
+{{/let}}
+```
+
+Like `let` in JavaScript and most functional languages, bindings can be
+"temporarily" overriden within a scope, and then "restored" once they
+pass out of scope:
+
+```handlebars
+{{let a=1 b=2}}
+a + b = {{sum a b}} <!-- 3 -->
+<div>
+  {{let a=5}}
+  a + b = {{sum a b}} <!-- 7 -->
+</div>
+a + b = {{sum a b}} <!-- 3 -->
+```
+
+Since this expands to (approximately):
+
+``` handlebars
+{{#let 1 2 as |a b|}}
+  a + b = {{sum a b}} <!-- 3 -->
+  <div>
+    {{#let 5 as |a|}}
+      a + b = {{sum a b}} <!-- 7 -->
+    {{/let}}
+  </div>
+  a + b = {{sum a b}} <!-- 3 -->
+{{/let}}
```

Comment:
I'm honestly worried that we'd start seeing templates like this in client projects :trollface: